NetKet/Graph/lattice_impl.hpp

```diff
@@ -0,0 +1,430 @@
+  for (int i = 0; i < ndim_; i++) {
+    if (basis_vectors_[i].size() != ndim_) {
+      ErrorMessage() << "Each element of basis_vectors must have ndim"
+                        "components. Dimensions are inconsistent. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  if (extent_.size() != ndim_) {
+    ErrorMessage() << "Extent must have ndim components. Dimension is "
+                      "inconsistent. Exit.\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // default case is a single atom in the unit cell, located at the origin
+  if (atoms_coord_.size() == 0) {
+    atoms_coord_.resize(1, std::vector<double>(ndim_, 0));
+  }
+
+  natoms_ = atoms_coord_.size();
+
+  for (int i = 0; i < natoms_; i++) {
+    if (atoms_coord_[i].size() != ndim_) {
+      ErrorMessage() << "Each element of atoms_coord must have ndim "
+                        "components. Dimension is inconsistent. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  // default case for pbc is true
+  if (pbc_.size() == 0) {
+    pbc_.resize(ndim_, true);
+  }
+
+  if (pbc_.size() != ndim_) {
+    ErrorMessage() << "Pbc must have ndim components. Dimension is "
+                      "inconsistent. Exit.\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // PATHOLOGIC 1: pbc(i) for extent(i)<=2 + extent(i)=0)
+  for (int i = 0; i < ndim_; i++) {
+    if (extent_[i] <= 2 and pbc_[i] == true) {
+      ErrorMessage() << "PBC are not allowed when extent<2. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+    if (extent_[i] == 0) {
+      ErrorMessage() << "Extent components must be >0. Exit.\n";
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  nlatticesites_ = 1;
+  for (int k = 0; k < ndim_; k++) nlatticesites_ *= extent_[k];
+
+  // PATHOLOGIC 2: nlatticesites_=0,1
+  if (nlatticesites_ <= 1) {
+    ErrorMessage()
+        << "A well-defined lattice must have at least 2 sites. Exit.\n";
+    exit(EXIT_FAILURE);
+  }
+
+  nsites_ = nlatticesites_ * natoms_;
+
+  for (int k = 0; k < natoms_; k++) {
+    for (int i = 0; i < nlatticesites_; i++) {
+      std::vector<int> n = Site2Vector(i);  // from site i to coefficents
+      std::vector<double> R =
+          Vector2Coord(n, k);  // initialize the R vector (coord of one site)
+      R_.push_back(R);
+    }
+  }
+
+  edges_ = BuildEdges();
+
+  symmetrytable_ = BuildSymmetryTable();
+
+  colors_.reserve(nlatticesites_);
+  for (auto const &edge : edges_) {
+    auto success = colors_.emplace(edge, 0).second;
+    static_cast<void>(success);  // Make everyone happy in the NDEBUG case
+    assert(success && "There should be no duplicate edges");
+  }
+
+  is_connected_ = ComputeConnected();
+  is_bipartite_ = ComputeBipartite();
+
+};  // namespace netket
+Lattice::~Lattice(){};
+
+// Get private member
+int Lattice::Ndim() const noexcept { return ndim_; };
+int Lattice::Nsites() const noexcept { return nsites_; };
+int Lattice::Size() const noexcept { return Nsites(); };
+
+std::vector<std::vector<double>> Lattice::Coordinates() const noexcept {

```

Comment:
You allocate memory using `std::allocator<...>`, which is a potentially throwing operation. Marking such functions `noexcept` may lead to crashes later on.

Reply:
fixed